python/sglang/srt/mem_cache/sparsity/algorithms/deepseek_nsa.py

```diff
@@ -0,0 +1,74 @@
+import logging
+from typing import TYPE_CHECKING, Any, Optional
+
+import nvtx
+import torch
+
+from sglang.srt.mem_cache.sparsity.algorithms.base_algorithm import (
+    BaseSparseAlgorithm,
+    SparseMode,
+)
+
+if TYPE_CHECKING:
+    from sglang.srt.layers.attention.nsa.nsa_indexer import Indexer
+    from sglang.srt.model_executor.forward_batch_info import ForwardBatch
+
+logger = logging.getLogger(__name__)
+
+
+class DeepSeekNSAAlgorithm(BaseSparseAlgorithm):
+    """Sparse attention algorithm for DeepSeek NSA."""
+
+    def __init__(self, config, device: torch.device, **kwargs):
+        super().__init__(config, device, **kwargs)
+        self.index_topk = getattr(config, "index_topk", 2048)
+        logger.info(f"DeepSeekNSAAlgorithm initialized: index_topk={self.index_topk}")
+
+    def get_sparse_mode(self) -> SparseMode:
+        return SparseMode.DEEPSEEK_TOKEN_WISE
+
+    @nvtx.annotate("DeepSeekNSAAlgorithm.retrieve_topk", color="green")
+    def retrieve_topk(
+        self,
+        queries: torch.Tensor,
+        layer_id: int,
+        req_pool_indices: torch.Tensor,
+        sparse_mask: torch.Tensor,
+        attn_metadata: Optional[Any],
+        **kwargs,
+    ) -> tuple:
+        indexer: Optional["Indexer"] = kwargs.get("indexer")
```

Comment:
should we have indexer as an member of the class?

Reply:
I'll take a try.